server/container/util-java8/src/main/java/org/apache/james/util/FunctionGenerator.java

```diff
@@ -0,0 +1,29 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.util;
+
+import java.util.function.Function;
+
+@FunctionalInterface
+public interface FunctionGenerator<A, R> extends Function<A, Function<R, R>> {
```

Comment:
The concerned objects are Functions that turns out to produce functions.
The idea of this class is to add availability to "chain" such objects.
You should think of this as Functions that generates functions. Hence my choice **FunctionGenerator**.

If you have better idea for naming, I listen. On my side, I have troubles finding names that better capture the itent I had...

For your code snipset, it is wrong, as my function is doing something with the result. Here is what inspection returns for "andThen" call : 

```
default <V> Function<T, V> andThen(Function<? super R, ? extends V> var1) {
        Objects.requireNonNull(var1);
        return (var2) -> {
            return var1.apply(this.apply(var2));
        };
    }
```
Maybe I can use **compose** here to make it easier to reson about it...